### ISSUE TITLE
Questions on some portions of your paper

### ISSUE THREAD

**Issue by ksachdeva** [FGiuliari/Trajectory-Transformer#7]
Hi,

Congratulations for creating the state of art in trajectory prediction and thanks for sharing the code as well.

I read your paper & have gone through the code and have few questions. I am fairly new both to the transformer and trajectory prediction domain itself.

Would appreciate if you could provide guidance here.

----

On page 12 (at the bottom and continues to page 13), you say -

> For example t-kth sample is missing, transformer will use the remaining t-7,...,t-k-1, t-k+1,..t .... to perform the prediction

Wouldn't this change the number of points needed for observation and hence would not work with the static input size required by the network. 

> Structural ability is absent in LSTM and RNN and they can not work with missing data and therefore Transformer is superior. 

You mention that one could use Linear Interpolation. I would imagine that one would do it for LSTM as well. 

----

On page 14 of your paper, you say -

> In general, we observed that LSTM generates trajectories way more regular than those predicted by TF, and this is certainly motivated by its unrolling, opposed to encoder-decoder architecture of TF

What does `regular` mean here ? If you could elaborate here. It seems that you are trying to talk about an advantage of LSTM here but I am not able to understand what exactly that is.

----

Is it possible to provide some numbers on how much time it takes typically to train TF network ? and what are the typical inference times ?

Regards & thanks
Kapil

**Comment by FGiuliari** [comment]
Hi Kapil,
I'm more than happy to answer any question you have about the paper.

#
>Wouldn't this change the number of points needed for observation and hence would not work with the static input size required by the network.

One of the cool things about the Transformer (and LSTM and all the other RNN methods) is that you do not have to have a fixed-sized sequence length as input. 
For the transformer all of the elements of the sequence are treated the same and share information during the self-attention phase. **Without** positional encoding the network would not even assume that there is an order for the sequence elements; since we **do** have an order we of course employ the use of a positional encoding.

#
>You mention that one could use Linear Interpolation. I would imagine that one would do it for LSTM as well.

Indeed one could attempt to use linear interpolation to mitigate the problem introduced by missing observation, and that's true for both Transformer and LSTM. We did not investigate deeply the linear interpolation route for neither of the models because we wanted to focus on a way to solve this task without relying on linear interpolation.

#
>What does regular mean here? If you could elaborate here. It seems that you are trying to talk about an advantage of LSTM here but I am not able to understand what exactly that is.

By regular we actually meant straight or linear; it will have to be fixed in the next revision of the paper because it is indeed confusing.

#
>Is it possible to provide some numbers on how much time it takes typically to train TF network ? and what are the typical inference times?

This is highly subjective to the kind of hardware you have and the dimension of the model, with standard dimensions on a 2080ti it takes 0.11 seconds to train a batch, and 0.25 seconds to test a batch.

#

I hope that clarified some doubts you had, if not feel free to ask.

**Comment by ksachdeva** [comment]
Many thanks @FGiuliari for taking the time to answer. Much appreciated. 

> One of the cool things about the Transformer (and LSTM and all the other RNN methods) is that you do not have to have a fixed-sized sequence length as input.
For the transformer all of the elements of the sequence are treated the same and share information during the self-attention phase. Without positional encoding the network would not even assume that there is an order for the sequence elements; since we do have an order we of course employ the use of a positional encoding.

Let's say that we have 8 observation points i.e. no missing data

These points are observed at following times
t_1, t_2, t_3, t_4, t_5, t_6, t_7, t_8

these points(embeddings) are combined with positional encoding. Thanks to positional encoding, the network learns the relative distance between observations and therefore have a notion of order of observations.

The part that is not clear to me is following -

Because of sensor issues let's say we could only observe 7 points and observation at t_5 is missing.

t_1, t_2, t_3, t_4, t_6, t_7, t_8

Based on your answer on the variable size input support of Transformer (& LSTM/RNN), we would now simply put 7 observations

The positional encoding will now be combined with 7 observations. Since the transformer (positional encoding in particular) does not really have notion of time but simply the relative distance/position wouldn't it assume the same distance between t_4 and t_6 at inference time as that t_2 and t_3 or t_3 and t_4 ? 

Phrasing the above question in another way, how would positional encoding help here in identifying the missing observation ?

**Comment by ksachdeva** [comment]
Hi @FGiuliari 

Few more questions and comments 

The paper does not talk about data prep pipeline (like using only 7 steps, standardization and normalization of the points etc).I have figured it out by debugging the code but may be in your next version/update it would be a good idea to have information on it.

Having a hard time understanding the lines used to prepare the source and target tensors masks just before feeding them to the model.

I have my comments (with KS prefix) inlines in the below snippet. Would appreciate if you could help clarify.

```
# KS : I passed a batch size of 2 to try it

for id_b,batch in enumerate(tr_dl):

            optim.optimizer.zero_grad()
 
            # KS - you are picking the last 2 entries in each input and you ignore the first
            # input because it is 0,0 
            inp=(batch['src'][:,1:,2:4].to(device)-mean.to(device))/std.to(device)

            # KS - you are ignoring the last entry in the target. Do not understand the reason for it.
            target=(batch['trg'][:,:-1,2:4].to(device)-mean.to(device))/std.to(device)
 
            # KS - target shape is [2, 11, 2] ... 11 and not 12 because you excluded the last entry
            # KS - you create target_c and its shape is [2,11,1] 
            target_c=torch.zeros((target.shape[0],target.shape[1],1)).to(device)
            # KS - after concate the shape is [2, 11, 3] , What does this third entry in the target used for ?
            target=torch.cat((target,target_c),-1)

            # KS - start of sequence is [2, 1, 3] shaped with values [0,0,1] 
            start_of_seq = torch.Tensor([0, 0, 1]).unsqueeze(0).unsqueeze(1).repeat(target.shape[0],1,1).to(device)

            # KS - decoder input is shaped [2,12,3] .... so you added above start of sequence 
            # to make it 12 steps. I am not sure about the reasons for this
            dec_inp = torch.cat((start_of_seq, target), 1)

            # KS - src_att is really src_mask and trg_att is really trg_mask
            src_att = torch.ones((inp.shape[0], 1,inp.shape[1])).to(device)

            # KS - src_att shape is [2,1,7] with all entries to 1 ... this most likely means that
            # we are not asking the transformer to ignore any points in the src_inp
 
            # KS - This generates a triangluar matrix with lower entries set to 1 (True). I think this is same as what is done in transformer with NLP but still would help some clarification in the context of trajectory
            trg_att=subsequent_mask(dec_inp.shape[1]).repeat(dec_inp.shape[0],1,1).to(device)

            pred=model(inp, dec_inp, src_att, trg_att)
```

Regards
Kapil

**Comment by FGiuliari** [comment]
> 
> 
> Hi @FGiuliari
> 
> Few more questions and comments
> 
> The paper does not talk about data prep pipeline (like using only 7 steps, standardization and normalization of the points etc).I have figured it out by debugging the code but may be in your next version/update it would be a good idea to have information on it.
> 
> Having a hard time understanding the lines used to prepare the source and target tensors masks just before feeding them to the model.
> 
> I have my comments (with KS prefix) inlines in the below snippet. Would appreciate if you could help clarify.
> 
> ```
> # KS : I passed a batch size of 2 to try it
> 
> for id_b,batch in enumerate(tr_dl):
> 
>             optim.optimizer.zero_grad()
>  
>             # KS - you are picking the last 2 entries in each input and you ignore the first
>             # input because it is 0,0 
>             inp=(batch['src'][:,1:,2:4].to(device)-mean.to(device))/std.to(device)
> 
>             # KS - you are ignoring the last entry in the target. Do not understand the reason for it.
>             target=(batch['trg'][:,:-1,2:4].to(device)-mean.to(device))/std.to(device)
>  
>             # KS - target shape is [2, 11, 2] ... 11 and not 12 because you excluded the last entry
>             # KS - you create target_c and its shape is [2,11,1] 
>             target_c=torch.zeros((target.shape[0],target.shape[1],1)).to(device)
>             # KS - after concate the shape is [2, 11, 3] , What does this third entry in the target used for ?
>             target=torch.cat((target,target_c),-1)
> 
>             # KS - start of sequence is [2, 1, 3] shaped with values [0,0,1] 
>             start_of_seq = torch.Tensor([0, 0, 1]).unsqueeze(0).unsqueeze(1).repeat(target.shape[0],1,1).to(device)
> 
>             # KS - decoder input is shaped [2,12,3] .... so you added above start of sequence 
>             # to make it 12 steps. I am not sure about the reasons for this
>             dec_inp = torch.cat((start_of_seq, target), 1)
> 
>             # KS - src_att is really src_mask and trg_att is really trg_mask
>             src_att = torch.ones((inp.shape[0], 1,inp.shape[1])).to(device)
> 
>             # KS - src_att shape is [2,1,7] with all entries to 1 ... this most likely means that
>             # we are not asking the transformer to ignore any points in the src_inp
>  
>             # KS - This generates a triangluar matrix with lower entries set to 1 (True). I think this is same as what is done in transformer with NLP but still would help some clarification in the context of trajectory
>             trg_att=subsequent_mask(dec_inp.shape[1]).repeat(dec_inp.shape[0],1,1).to(device)
> 
>             pred=model(inp, dec_inp, src_att, trg_att)
> ```
> 
> Regards
> Kapil

Hi, 
I will clarify these points now and answer your questions about the paper later when i'll have a little more time.

>KS - you are picking the last 2 entries in each input and you ignore the first input because it is 0,0 

yes, we work with velocity vectors instead of absolute positions since we found that it works better that way.
both batch['src'] and batch['trg] are N x L x 4 matrices where N is the number of trajectories, L is the number of observed intervals or To_predict intervals, and 4 are [x,y, speed_x,speed_y] where speed_x and speed_y are a simple difference between x (or y) and x_t-1 (or y_t-1).
Since we work on differences we work on obs-1 intervals.

>KS - you are ignoring the last entry in the target. Do not understand the reason for it.

those are the input to the decoder part since we train with forced teaching.
so you need to have preds-1 inputs, because preds-1 will give the last prediction. 



>KS - after concate the shape is [2, 11, 3] , What does this third entry in the target used for ?

the decoder needs to have a "start of sequence" token from which to make the first prediction. What i did was have a special [0,0,1] input as that token. The third feature is used to let the model understand which is the start_of_seq token and what are previous prediction/inputs; it is 1 for the start_of_seq, and 0 when the input element is a velocity vector.

>KS - decoder input is shaped [2,12,3] .... so you added above start of sequence to make it 12 steps. I am not sure about the reasons for this.

Refer to the above explanation. Maybe there is a better way to encode this, but that was the easiest to implement and seemed to work fine.

>KS - src_att is really src_mask and trg_att is really trg_mask

correct, those two variables are used to calculate the masked self_attention


>KS - src_att shape is [2,1,7] with all entries to 1 ... this most likely means that we are not asking the transformer to ignore any points in the src_inp

Correct, in the encoder all the input elements can attend to all the other input elements.

>KS - This generates a triangular matrix with lower entries set to 1 (True). I think this is same as what is done in the transformer with NLP but still would help some clarification in the context of trajectory.

Since when doing inference we will work in an autoregressive manner we need to also have this behaviour when training.

So we will need to mask decoder inputs that will follow the current one.
So the start_of_sequence attention will be calculated only on itself and not on all the other inputs that follow it.


hoped that cleared some of the workings of the code.

#

Since you seemed really interested in transformer models I would also suggest to look at this blog post https://jalammar.github.io/illustrated-transformer/
and this code https://nlp.seas.harvard.edu/2018/04/03/attention.html

They really helped me when I was starting working with them and I wanted to share them in case you didn't already find them

br,
Francesco

**Comment by ksachdeva** [comment]
Thanks @FGiuliari  for the answers, they are very helpful. 

Indeed one of my reasons to explore your paper is to learn more about Transformer and its applications to areas other than NLP. Your work is a perfect example of that. I have been reading more on it in last few days and did come across the links you have provided. They are really good. 

I understand things better by implementing them so I started to implement/port your paper to Tensorflow 2.0. Yesterday night I had all the components in place (data pipeline, various custom layers and models, loss function) and then started the training.

Unfortunately at present, I am getting NaN after few epochs so there must be something wrong somewhere and am trying to figure out how to debug it. It is always difficult to NaN related issues that do not occur systematically when training. 

I have my repo in private for now. If I can make it train successfully I would make it public.

Regards
Kapil

**Comment by ksachdeva** [comment]
Hi @FGiuliari 

Some good news. I managed to resolve the NaN issue (hopefully) in my tensorflow implementation as haven't seen it now for few hundred epochs. [Still have to work on the validation and test part]

I am also able to run your implementation on google colab after few rearrangements of training script. If you want I can submit my changes such that it can be easy to run on google colab.

I ran your model for 400 epochs the results are 

Validation accuracy on `eth` is 
`MAD=0.44007973403116074, FAD=0.9562298671734019`

Test accuracy on `eth` is
`MAD=1.1035755287112063, FAD=2.290178309514566`

Is this corresponds to Table 3 in the paper (last entry) [1.03/2.10] ? . I understand that it is different from your paper because I just ran the final model (400th epoch one) and it is quite possible there may be one in the middle with better accuracy or the one that matches yours.

Initially I was confused about all these datasets but it seems that you train on `eth` and test on `eth`. You then train a different model on `univ` and then test on `univ`. Is this correct understanding ? Does this mean that model trained on one dataset does not generalize on other datasets ?

Regards
Kapil

**Comment by FGiuliari** [comment]
Hi Kapil,
yes the results should correspond to table 3.

On the training protocol is kind of the other way around.
For trajectory forecasting on eth+ucy the standard protocol is to follow a leave one out approach, where you train on all the remaining scenes and test on the chosen one.

In the dataset folder the data are already subdivided so for example "eth/test" contains the testing data, which are trajectories from the eth scene; "eth/train" and "eth/val" contains the training and validation data, which are trajectories for the remaining 4 scenes.

So you need a model for each scene, but not because you train on the data on that scene and it doesn't generalize, but because the training data contains what would be testing trajectories from other scenes.

**Comment by ksachdeva** [comment]
Thanks @FGiuliari it makes more sense now how the datasets are organized in your repository. 

I downloaded the trajnet challenge dataset as well and there the organization is bit different. They have train and test folders which have subfolders.

For trajnet challenge, does one the follow the same protocol as above (leave one out) or one is supposed to train one model that uses all the .txt in the train (and its subfolders)and test on the files in the test folder albeit report the results on individual files in test folder

**Comment by jantonio07** [comment]
Hi!

Congratulations, @FGiuliari, your work is very important in many senses.

I have a question about something in the paper and I think this is a good space to ask. Sorry if inappropriate. In Table 2 there is a column for Trajectron++, but the results are different from the original paper. Did you run their code again?

By the way, could you tell me the number of parameters the model TFq has? It would be good to know.

Best regards,
Jose

**Comment by FGiuliari** [comment]
Hi Jantonio,
The numbers are different because we are referencing their revision 1, which was the most recent one at the time of the writing of the paper. You can find the numbers in table 5 here https://arxiv.org/pdf/2001.03093v1.pdf

As for the number of parameters we used the same architectures was in the original "attention is all you need paper"; we had some minor changes but they should not change the order of magnitude of the parameters by much.

Best regards,
Francesco

**Comment by manybodycpa** [comment]
Dear Francesco,

Thanks for providing your code here. Following up on the question regarding the third column (which contains the start token), I was wondering if it is completely obvious that the values in this column should contribute to the loss during training? In principle, only 0 and 1 are allowed, whereas imperfect training will give intermediate values. Have you ever thought about how this compares to the use of special tokens in NLP?

Best wishes,

Martin

**Comment by FindTheAnswer** [comment]
> Many thanks @FGiuliari for taking the time to answer. Much appreciated.
> 
> > One of the cool things about the Transformer (and LSTM and all the other RNN methods) is that you do not have to have a fixed-sized sequence length as input.
> > For the transformer all of the elements of the sequence are treated the same and share information during the self-attention phase. Without positional encoding the network would not even assume that there is an order for the sequence elements; since we do have an order we of course employ the use of a positional encoding.
> 
> Let's say that we have 8 observation points i.e. no missing data
> 
> These points are observed at following times
> t_1, t_2, t_3, t_4, t_5, t_6, t_7, t_8
> 
> these points(embeddings) are combined with positional encoding. Thanks to positional encoding, the network learns the relative distance between observations and therefore have a notion of order of observations.
> 
> The part that is not clear to me is following -
> 
> Because of sensor issues let's say we could only observe 7 points and observation at t_5 is missing.
> 
> t_1, t_2, t_3, t_4, t_6, t_7, t_8
> 
> Based on your answer on the variable size input support of Transformer (& LSTM/RNN), we would now simply put 7 observations
> 
> The positional encoding will now be combined with 7 observations. Since the transformer (positional encoding in particular) does not really have notion of time but simply the relative distance/position wouldn't it assume the same distance between t_4 and t_6 at inference time as that t_2 and t_3 or t_3 and t_4 ?
> 
> Phrasing the above question in another way, how would positional encoding help here in identifying the missing observation ?

Hi, have you got this problem clear? I have same question with you.
